File: otherImpl/c#/libComputeDuck/Chunk.cs
using System;
using System.Collections.Generic;

namespace ComputeDuck
{
    using OpCodes = List<int>;
    public enum OpCode
    {
        OP_CONSTANT = 0,
        OP_ADD,
        OP_SUB,
        OP_MUL,
        OP_DIV,
        OP_EQUAL,
        OP_GREATER,
        OP_LESS,
        OP_NOT,
        OP_MINUS,
        OP_AND,
        OP_OR,
        OP_BIT_AND,
        OP_BIT_OR,
        OP_BIT_NOT,
        OP_BIT_XOR,
        OP_JUMP_IF_FALSE,
        OP_JUMP,
        OP_DEF_GLOBAL,
        OP_SET_GLOBAL,
        OP_GET_GLOBAL,
        OP_DEF_LOCAL,
        OP_SET_LOCAL,
        OP_GET_LOCAL,
        OP_ARRAY,
        OP_GET_INDEX,
        OP_SET_INDEX,
        OP_FUNCTION_CALL,
        OP_RETURN,
        OP_GET_BUILTIN,
        OP_STRUCT,
        OP_GET_STRUCT,
        OP_SET_STRUCT,
        OP_REF_GLOBAL,
        OP_REF_LOCAL,
        OP_REF_INDEX_GLOBAL,
        OP_REF_INDEX_LOCAL,
        OP_SP_OFFSET,
        OP_DLL_IMPORT,
    };

    public class Chunk
    {
        public OpCodes opCodes;
        public List<Object> constants;

        public Chunk()
        {
            this.opCodes = new OpCodes();
            this.constants = new List<Object>();
        }
        public Chunk(OpCodes opCodes, List<Object> constants)
        {
            this.opCodes = opCodes;
            this.constants = constants;
        }

        public override string ToString()
        {
            string result = OpCodeStringify(opCodes);
            for (int i = 0; i < constants.Count; ++i)
                if (constants[i].type == ObjectType.FUNCTION)
                    result += ((FunctionObject)constants[i]).ToStringWithChunk();

            return result;
        }

        private string OpCodeStringify(OpCodes opCodes)
        {
            string result = "";
            for (int i = 0; i < opCodes.Count; ++i)
            {
                switch (opCodes[i])
                {
                    case (int)OpCode.OP_CONSTANT:
                        result += string.Format("{0}\tOP_CONSTANT\t'{1}'\t\n", i.ToString().PadLeft(8), constants[opCodes[++i]].ToString());
                        break;
                    case (int)OpCode.OP_ADD:
                        result += string.Format("{0}\tOP_ADD\n", i.ToString().PadLeft(8));
                        break;
                    case (int)OpCode.OP_SUB:
                        result += string.Format("{0}\tOP_SUB\n", i.ToString().PadLeft(8));
                        break;
                    case (int)OpCode.OP_MUL:
                        result += string.Format("{0}\tOP_MUL\n", i.ToString().PadLeft(8));
                        break;
                    case (int)OpCode.OP_DIV:
                        result += string.Format("{0}\tOP_DIV\n", i.ToString().PadLeft(8));
                        break;
                    case (int)OpCode.OP_LESS:
                        result += string.Format("{0}\tOP_LESS\n", i.ToString().PadLeft(8));
                        break;
                    case (int)OpCode.OP_GREATER:
                        result += string.Format("{0}\tOP_GREATER\n", i.ToString().PadLeft(8));
                        break;
                    case (int)OpCode.OP_NOT:
                        result += string.Format("{0}\tOP_NOT\n", i.ToString().PadLeft(8));
                        break;
                    case (int)OpCode.OP_MINUS:
                        result += string.Format("{0}\tOP_MINUS\n", i.ToString().PadLeft(8));
                        break;
                    case (int)OpCode.OP_EQUAL:
                        result += string.Format("{0}\tOP_EQUAL\n", i.ToString().PadLeft(8));
                        break;
                    case (int)OpCode.OP_AND:
                        result += string.Format("{0}\tOP_AND\n", i.ToString().PadLeft(8));
                        break;
                    case (int)OpCode.OP_OR:
                        result += string.Format("{0}\tOP_OR\n", i.ToString().PadLeft(8));
                        break;
                    case (int)OpCode.OP_BIT_AND:
                        result += string.Format("{0}\tOP_BIT_AND\n", i.ToString().PadLeft(8));
                        break;
                    case (int)OpCode.OP_BIT_OR:
                        result += string.Format("{0}\tOP_BIT_OR\n", i.ToString().PadLeft(8));
                        break;
                    case (int)OpCode.OP_BIT_XOR:
                        result += string.Format("{0}\tOP_BIT_XOR\n", i.ToString().PadLeft(8));
                        break;
                    case (int)OpCode.OP_BIT_NOT:
                        result += string.Format("{0}\tOP_BIT_NOT\n", i.ToString().PadLeft(8));
                        break;
                    case (int)OpCode.OP_ARRAY:
                        result += string.Format("{0}\tOP_ARRAY\t{1}\n", i.ToString().PadLeft(8), opCodes[++i]);
                        break;
                    case (int)OpCode.OP_GET_INDEX:
                        result += string.Format("{0}\tOP_GET_INDEX\n", i.ToString().PadLeft(8));
                        break;
                    case (int)OpCode.OP_SET_INDEX:
                        result += string.Format("{0}\tOP_SET_INDEX\n", i.ToString().PadLeft(8));
                        break;
                    case (int)OpCode.OP_JUMP:
                        result += string.Format("{0}\tOP_JUMP\t{1}\n", i.ToString().PadLeft(8), opCodes[++i]);
                        break;
                    case (int)OpCode.OP_JUMP_IF_FALSE:
                        result += string.Format("{0}\tOP_JUMP_IF_FALSE\t{1}\n", i.ToString().PadLeft(8), opCodes[++i]);
                        break;
                    case (int)OpCode.OP_RETURN:
                        result += string.Format("{0}\tOP_RETURN\t{1}\n", i.ToString().PadLeft(8), opCodes[++i]);
                        break;
                    case (int)OpCode.OP_DEF_GLOBAL:
                        result += string.Format("{0}\tOP_DEF_GLOBAL\t{1}\n", i.ToString().PadLeft(8), opCodes[++i]);
                        break;
                    case (int)OpCode.OP_SET_GLOBAL:
                        result += string.Format("{0}\tOP_SET_GLOBAL\t{1}\n", i.ToString().PadLeft(8), opCodes[++i]);
                        break;
                    case (int)OpCode.OP_GET_GLOBAL:
                        result += string.Format("{0}\tOP_GET_GLOBAL\t{1}\n", i.ToString().PadLeft(8), opCodes[++i]);
                        break;
                    case (int)OpCode.OP_DEF_LOCAL:
                        result += string.Format("{0}\tOP_DEF_LOCAL\t{1}\t{2}\n", i.ToString().PadLeft(8), opCodes[++i], opCodes[++i]);
                        break;
                    case (int)OpCode.OP_SET_LOCAL:
                        result += string.Format("{0}\tOP_SET_LOCAL\t{1}\t{2}\t{3}\n", i.ToString().PadLeft(8), opCodes[++i], opCodes[++i], opCodes[++i]);
                        break;
                    case (int)OpCode.OP_GET_LOCAL:
                        result += string.Format("{0}\tOP_GET_LOCAL\t{1}\t{2}\t{3}\n", i.ToString().PadLeft(8), opCodes[++i], opCodes[++i], opCodes[++i]);
                        break;
                    case (int)OpCode.OP_FUNCTION_CALL:
                        result += string.Format("{0}\tOP_FUNCTION_CALL\t{1}\n", i.ToString().PadLeft(8), opCodes[++i]);
                        break;
                    case (int)OpCode.OP_GET_BUILTIN:
                        result += string.Format("{0}\tOP_GET_BUILTIN\n", i.ToString().PadLeft(8));
                        break;
                    case (int)OpCode.OP_STRUCT:
                        result += string.Format("{0}\tOP_STRUCT\t{1}\n", i.ToString().PadLeft(8), opCodes[++i]);
                        break;
                    case (int)OpCode.OP_GET_STRUCT:
                        result += string.Format("{0}\tOP_GET_STRUCT\n", i.ToString().PadLeft(8));
                        break;
                    case (int)OpCode.OP_SET_STRUCT:
                        result += string.Format("{0}\tOP_SET_STRUCT\n", i.ToString().PadLeft(8));
                        break;
                    case (int)OpCode.OP_REF_GLOBAL:
                        result += string.Format("{0}\tOP_REF_GLOBAL\t{1}\n", i.ToString().PadLeft(8), opCodes[++i]);
                        break;
                    case (int)OpCode.OP_REF_LOCAL:
                        result += string.Format("{0}\tOP_REF_LOCAL\t{1}\t{2}\t{3}\n", i.ToString().PadLeft(8), opCodes[++i], opCodes[++i], opCodes[++i]);
                        break;
                    case (int)OpCode.OP_REF_INDEX_GLOBAL:
                        result += string.Format("{0}\tOP_REF_INDEX_GLOBAL\t{1}\n", i.ToString().PadLeft(8), opCodes[++i]);
                        break;
                    case (int)OpCode.OP_REF_INDEX_LOCAL:
                        result += string.Format("{0}\tOP_REF_INDEX_LOCAL\t{1}\t{2}\t{3}\n", i.ToString().PadLeft(8), opCodes[++i], opCodes[++i], opCodes[++i]);
                        break;
                    case (int)OpCode.OP_SP_OFFSET:
                        result += string.Format("{0}\tOP_SP_OFFSET\t{1}\n", i.ToString().PadLeft(8), opCodes[++i]);
                        break;
                    case (int)OpCode.OP_DLL_IMPORT:
                        result += string.Format("{0}\tOP_DLL_IMPORT\n", i.ToString().PadLeft(8));
                        break;
                    default:
                        break;
                }
            }

            return result;
        }

    }

}

File: otherImpl/python/Chunk.py
from enum import IntEnum
from Object import *
class OpCode(IntEnum):
    OP_CONSTANT=0,
    OP_ADD=1,
    OP_SUB=2,
    OP_MUL=3,
    OP_DIV=4,
    OP_EQUAL=5,
    OP_GREATER=6,
    OP_LESS=7,
    OP_NOT=8,
    OP_MINUS=9,
    OP_AND=10,
    OP_OR=11,
    OP_BIT_AND=12,
    OP_BIT_OR=13,
    OP_BIT_NOT=14,
    OP_BIT_XOR=15,
    OP_JUMP_IF_FALSE=16,
    OP_JUMP=17,
    OP_DEF_GLOBAL=18,
    OP_SET_GLOBAL=19,
    OP_GET_GLOBAL=20,
    OP_DEF_LOCAL=21,
    OP_SET_LOCAL=22,
    OP_GET_LOCAL=23,
    OP_ARRAY=24,
    OP_GET_INDEX=25,
    OP_SET_INDEX=26,
    OP_FUNCTION_CALL=27,
    OP_RETURN=28,
    OP_GET_BUILTIN=29,
    OP_STRUCT=30,
    OP_GET_STRUCT=31,
    OP_SET_STRUCT=32,
    OP_REF_GLOBAL=33,
    OP_REF_LOCAL=34,
    OP_REF_INDEX_GLOBAL=35,
    OP_REF_INDEX_LOCAL=36,
    OP_DLL_IMPORT=37

class Chunk:
    opCodes:list[int]
    constants:list[Object]

    def __init__(self,opCodes:list[int]=[],constants:list[Object]=[]) -> None:
        if opCodes!=[]:
            self.opCodes=opCodes
        else:
            self.opCodes=[]
            
        if constants!=[]:
            self.constants=constants
        else:
            self.constants=[]

    def __str__(self)->str:
        result = self.__opcode_stringify(self.opCodes)
        for i in range(0,len(self.constants)):
            if self.constants[i].type==ObjectType.FUNCTION:
                result+=self.constants[i].str_with_chunk()

        return result

    def __opcode_stringify(self,opcodes:list[int])->str:
        result=""
        i = 0
        while i < (len(opcodes)):
            if opcodes[i]==OpCode.OP_CONSTANT:
                pos=opcodes[i+1]
                result+=("%8d\tOP_CONSTANT\t%d\t'%s'\n" % (i,pos,self.constants[pos]))
                i=i+1
            elif opcodes[i]==OpCode.OP_ADD:
                result+=("%8d\tOP_ADD\n" % (i))
            elif opcodes[i]==OpCode.OP_SUB:
                result+=("%8d\tOP_SUB\n" % (i))
            elif opcodes[i]==OpCode.OP_MUL:
                result+=("%8d\tOP_MUL\n" % (i))
            elif opcodes[i]==OpCode.OP_DIV:
                result+=("%8d\tOP_DIV\n" % (i))
            elif opcodes[i]==OpCode.OP_LESS:
                result+=("%8d\tOP_LESS\n" % (i))
            elif opcodes[i]==OpCode.OP_GREATER:
                result+=("%8d\tOP_GREATER\n" % (i))
            elif opcodes[i]==OpCode.OP_MINUS:
                result+=("%8d\tOP_MINUS\n" % (i))
            elif opcodes[i]==OpCode.OP_EQUAL:
                result+=("%8d\tOP_EQUAL\n" % (i))
            elif opcodes[i]==OpCode.OP_AND:
                result+=("%8d\tOP_AND\n" % (i))
            elif opcodes[i]==OpCode.OP_OR:
                result+=("%8d\tOP_OR\n" % (i))
            elif opcodes[i]==OpCode.OP_NOT:
                result+=("%8d\tOP_NOT\n" % (i))
            elif opcodes[i]==OpCode.OP_BIT_AND:
                result+=("%8d\tOP_BIT_AND\n" % (i))
            elif opcodes[i]==OpCode.OP_BIT_OR:
                result+=("%8d\tOP_BIT_OR\n" % (i))
            elif opcodes[i]==OpCode.OP_BIT_XOR:
                result+=("%8d\tOP_BIT_XOR\n" % (i))
            elif opcodes[i]==OpCode.OP_BIT_NOT:
                result+=("%8d\tOP_BIT_NOT\n" % (i))
            elif opcodes[i]==OpCode.OP_ARRAY:
                count=opcodes[i+1]
                result+=("%8d\tOP_ARRAY\t%d\n" % (i,count))
                i=i+1
            elif opcodes[i]==OpCode.OP_GET_INDEX:
                result+=("%8d\tOP_GET_INDEX\n" % (i))
            elif opcodes[i]==OpCode.OP_SET_INDEX:
                result+=("%8d\tOP_SET_INDEX\n" % (i))
            elif opcodes[i]==OpCode.OP_JUMP:
                address=opcodes[i+1]
                result+=("%8d\tOP_JUMP\t%d\n" % (i,address))
                i=i+1
            elif opcodes[i]==OpCode.OP_JUMP_IF_FALSE:
                address=opcodes[i+1]
                result+=("%8d\tOP_JUMP_IF_FALSE\t%d\n" % (i,address))
                i=i+1
            elif opcodes[i]==OpCode.OP_RETURN:
                count=opcodes[i+1]
                result+=("%8d\tOP_RETURN\t%d\n" % (i,count))
                i=i+1
            elif opcodes[i]==OpCode.OP_DEF_GLOBAL:
                pos=opcodes[i+1]
                result+=("%8d\tOP_DEF_GLOBAL\t%d\n" % (i,pos))
                i=i+1
            elif opcodes[i]==OpCode.OP_SET_GLOBAL:
                pos=opcodes[i+1]
                result+=("%8d\tOP_SET_GLOBAL\t%d\n" % (i,pos))
                i=i+1
            elif opcodes[i]==OpCode.OP_GET_GLOBAL:
                pos=opcodes[i+1]
                result+=("%8d\tOP_GET_GLOBAL\t%d\n" % (i,pos))
                i=i+1
            elif opcodes[i]==OpCode.OP_DEF_LOCAL:
                scopeDepth=opcodes[i+1]
                index=opcodes[i+2]
                result+=("%8d\tOP_DEF_LOCAL\t%d\t%d\n" % (i,scopeDepth,index))
                i=i+2
            elif opcodes[i]==OpCode.OP_SET_LOCAL:
                scopeDepth=opcodes[i+1]
                index=opcodes[i+2]
                isUpValue=opcodes[i+3]
                result+=("%8d\tOP_SET_LOCAL\t%d\t%d\t%d\n" % (i,scopeDepth,index,isUpValue))
                i=i+3
            elif opcodes[i]==OpCode.OP_GET_LOCAL:
                scopeDepth=opcodes[i+1]
                index=opcodes[i+2]
                isUpValue=opcodes[i+3]
                result+=("%8d\tOP_GET_LOCAL\t%d\t%d\t%d\n" % (i,scopeDepth,index,isUpValue))
                i=i+3
            elif opcodes[i]==OpCode.OP_FUNCTION_CALL:
                argCount=opcodes[i+1]
                result+=("%8d\tOP_FUNCTION_CALL\t%d\n" % (i,argCount))
                i=i+1
            elif opcodes[i]==OpCode.OP_GET_BUILTIN:
                result+=("%8d\tOP_GET_BUILTIN\n" % i)
            elif opcodes[i]==OpCode.OP_STRUCT:
                memberCount=opcodes[i+1]
                result+=("%8d\tOP_STRUCT\t%d\n" % (i,memberCount))
                i+=1
            elif opcodes[i]==OpCode.OP_GET_STRUCT:
                result+=("%8d\tOP_GET_STRUCT\n" % (i))
            elif opcodes[i]==OpCode.OP_SET_STRUCT:
                result+=("%8d\tOP_SET_STRUCT\n" % (i))
            elif opcodes[i]==OpCode.OP_REF_GLOBAL:
                idx=opcodes[i+1]
                result+=("%8d\tOP_REF_GLOBAL\t%d\n" % (i,idx))
                i=i+1
            elif opcodes[i]==OpCode.OP_REF_LOCAL:
                scopeDepth=opcodes[i+1]
                index=opcodes[i+2]
                isUpValue=opcodes[i+3]
                result+=("%8d\tOP_REF_LOCAL\t%d\t%d\t%d\n" % (i,scopeDepth,index,isUpValue))
                i=i+3
            elif opcodes[i]==OpCode.OP_REF_INDEX_GLOBAL:
                pos=opcodes[i+1]
                result+=("%8d\tOP_REF_INDEX_GLOBAL\t%d\n" % (i,pos))
                i=i+1
            elif opcodes[i]==OpCode.OP_REF_INDEX_LOCAL:
                scopeDepth=opcodes[i+1]
                index=opcodes[i+2]
                isUpValue=opcodes[i+3]
                result+=("%8d\tOP_REF_INDEX_LOCAL\t%d\t%d\t%d\n" % (i,scopeDepth,index,isUpValue))
                i=i+3
            elif opcodes[i]==OpCode.OP_DLL_IMPORT:
                result+=("%8d\tOP_DLL_IMPORT\n" % (i))
            i=i+1
        return result